src/lib/serializer.ts

```diff
@@ -0,0 +1,253 @@
+  private static wallEncodeMap = new Map([
+    ['line-empty', 0b11],
+    ['line-wall', 0b01],
+    ['line-half', 0b10],
+  ]);
+
+  private static bitsPerWall = 2;
+  private static wallsPerByte = 3;
+
+  // Extracts the ith wall from an encoding in the characterUnMap
+  private static extractWallAtIndex(wallEncoding: number, i: number) {
+    const extractor = (1 << Serializer.bitsPerWall) - 1; // 0b11
+    const shift = i * Serializer.bitsPerWall;
+    const wallRepresentation = (wallEncoding >> shift) & extractor;
+    return wallRepresentation;
+  }
+
+  // Extracts a list of walls from a given wall encoding
+  private static extractWalls(wallEncoding: number) {
+    return [...Array(Serializer.wallsPerByte)].map((_, i) => {
+      return Serializer.extractWallAtIndex(wallEncoding, i);
+    });
+  }
+
+  private static packWalls(walls: number[]) {
+    return walls.reduce((previous, current, i) => {
+      const shift = Serializer.bitsPerWall * i;
+      return previous | (current << shift);
+    });
+  }
+
+  private static serializeWallsArray(elements: (SquareType | WallType)[]) {
+    const binaryList = elements
+      .filter((element) => 'className' in element)
+      .map((element) => {
+        const className = (element as WallType).className;
+        const wallEncoding = Serializer.wallEncodeMap.get(className);
+        if (wallEncoding === undefined) {
+          throw new URIError(
+            `Invalid className ${className}: cannot encode walls`,
+          );
+        } else {
+          return wallEncoding;
+        }
+      });
+
+    const hexString = Utils.chunk(binaryList, Serializer.wallsPerByte)
+      .map((walls) => Serializer.packWalls(walls))
+      .map((num) => {
+        const wallRepresentation = Serializer.characterMap.get(num);
+        if (wallRepresentation === undefined) {
+          throw new URIError(
+            `Invalid characterMap character, cannot encode walls: ${num}`,
+          );
+        } else {
+          return wallRepresentation;
+        }
+      })
+      .join('');
+
+    return hexString;
+  }
+
+  // Serializes walls as a packing of 3 binary numbers as a character representing digits between 0 and 64
+  static serializeLayoutWalls(layout: Layout) {
+    const walls = layout.layout
+      .map((row, i) => {
+        return row.filter((_, j) => i % 2 === 0 || j % 2 === 0); // remove corner walls
+      })
+      .flat();
+    return Serializer.serializeWallsArray(walls);
+  }
+
+  static encodeLayoutString(layout: Layout) {
+    let layoutString = `v2 ${layout.height}x${layout.width} `;
+    const numVerticalElements = layout.height * 2 - 1;
+    const numHorizontalElements = layout.width * 2 - 1;
+    for (let i = 0; i < numVerticalElements; i++) {
+      for (let j = 0; j < numHorizontalElements; j++) {
+        const element = layout.layout[i][j];
+        const notCornerWall = i % 2 === 0 || j % 2 === 0;
+        const isWall = 'className' in element;
```

Comment:
`instanceof`, as above; however, I would just do `isSquare` (where `i % 2 === 0 && j % 2 === 0`) to avoid confusion